core/storage/question/gae_models.py

```diff
@@ -365,51 +362,67 @@ def get_question_skill_links_based_on_difficulty_equidistributed_by_skill(
             # We fetch more questions here in order to try and ensure that the
             # eventual number of returned questions is sufficient to meet the
             # number requested, even after deduplication.
-            new_question_skill_link_models = (
-                equal_questions_query.fetch(question_count_per_skill * 2))
+            new_question_skill_link_models = equal_questions_query.fetch()
```

Comment:
Why did you remove the extra fetch of questions? We still have to ensure that we remove duplicates right? We should first remove duplicates and then do the randomization

Reply:
By removing the count I am fetching all the questions that match the difficulty(according to your suggestion) and randomly picking out of them. So essentially I am fetching more than question_count_per_skill * 2.

Reply:
Oh sorry I misunderstood, but I still don't understand why we are fetching all instead of double. @sophiewu6 had made the fetch to double instead of all because the performance drops significantly if we iterate over all records. @sophiewu6 do you have any thoughts?

Reply:
Fetching all questions can improve the randomness of selecting questions. However, there could be a large number of questions for one skill. In that case, fetching all questions could @rohitkatlaa Have you evaluated its performance before and after your change?

Reply:
I haven't been following this PR, but I saw @sophiewu6's comment and want to highlight that you absolutely cannot fetch "all" questions. There needs to be an upper bound because there may be thousands of questions.

Reply:
@sophiewu6  I haven't evaluated it's performance but what can be done is to fetch the questions with question id greater than a particular random id.(this was the method I used earlier)

Reply:
@seanlip @sophiewu6  Just to be clear I am only fetching all the questions within a particular skill_id with particular difficulty, and I think that is going to be limited. And the earlier method of using a random_id to fetch questions with question_id greater than the random_id cannot be used in this [function](https://github.com/oppia/oppia/blob/8d8689bc45a75d5f89025ff216a2eccea1c9c2e9/core/storage/question/gae_models.py#L327) because google [datastore](https://cloud.google.com/appengine/docs/standard/python/datastore/query-restrictions#inequality_filters_are_limited_to_at_most_one_property) allows only one inequality filter at a time.

Reply:
There may easily be > 100 questions for a particular skill ID with a particular difficulty. So this method still doesn't work.

Perhaps you can use a random offset? I.e. generate the offset number for the query randomly?

Reply:
As I have mentioned in the earlier comment it is not possible to use two inequality filters(difficulty and random offset) at the same time.

Reply:
I also noticed that problem last year. I have made a google doc for possible solutions. Maybe you can take a look? https://docs.google.com/document/d/1nghfz229A65qQGen57kwQrHlQ4qQ6nFMD0_bqdVrj0A/edit?usp=sharing

Reply:
I had a look in the doc, both the solutions requires two inequality filters(difficulty and random id) for this [function](https://github.com/oppia/oppia/blob/8d8689bc45a75d5f89025ff216a2eccea1c9c2e9/core/storage/question/gae_models.py#L327).

Reply:
Why does the random offset need to be an inequality filter? I was talking about the offset arg in https://cloud.google.com/appengine/docs/standard/python/datastore/gqlqueryclass#GqlQuery_fetch

Reply:
I am sorry I didn't understand what offset meant. I think the offset can be used to fetch the random results.

Reply:
I don't understand your response, sorry. Are you still stuck or does the clarification help?

Reply:
I am sorry earlier I didn't understand the meaning of offset. I think the offset can be used to fetch random results.

Reply:
Hi @rohitkatlaa, please do not delete earlier responses. If you do, then it becomes hard to follow the conversation.

Could you please clarify whether you are still stuck, or whether the previous clarification helps?

Reply:
I understood the concept and I pushed the changes. PTAL